example/advection/advection.cpp

```diff
@@ -0,0 +1,399 @@
+}
+
+void MeshBlock::ProblemGenerator(ParameterInput *pin) {
+  Container<Real>& rc = real_containers.Get();
+  CellVariable<Real>& q = rc.Get("advected");
+
+  for (int k=0; k<ncells3; k++) {
+    for (int j=0; j<ncells2; j++) {
+      for (int i=0; i<ncells1; i++) {
+        Real rsq = std::pow(pcoord->x1v(i),2) + std::pow(pcoord->x2v(j),2);
+        q(k,j,i) = (rsq < 0.15*0.15 ? 1.0 : 0.0);
+      }
+    }
+  }
+}
+
+void ParthenonManager::SetFillDerivedFunctions() {
+  parthenon::FillDerivedVariables::SetFillDerivedFunctions(Advection::PreFill, Advection::PostFill);
+}
+
+} // namespace parthenon
+
+// *************************************************//
+// define the "physics" package Advect, which      *//
+// includes defining various functions that control*//
+// how parthenon functions and any tasks needed to *//
+// implement the "physics"                         *//
+// *************************************************//
+namespace Advection {
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+  auto pkg = std::make_shared<StateDescriptor>("Advection");
+  Params& params = pkg->AllParams();
+
+  Real cfl = pin->GetOrAddReal("Advection", "cfl", 0.45);
+  params.Add("cfl", cfl);
+  Real vx  = pin->GetOrAddReal("Advection", "vx", 1.0);
+  params.Add("vx", vx);
+  Real vy  = pin->GetOrAddReal("Advection", "vy", 1.0);
+  params.Add("vy", vy);
+  Real refine_tol  = pin->GetOrAddReal("Advection", "refine_tol", 0.3);
+  params.Add("refine_tol", refine_tol);
+  Real derefine_tol  = pin->GetOrAddReal("Advection", "derefine_tol", 0.03);
+  params.Add("derefine_tol", derefine_tol);
+
+  std::string field_name = "advected";
+  Metadata m({Metadata::Cell, Metadata::Independent, Metadata::Graphics, Metadata::FillGhost});
+  pkg->AddField(field_name, m);
+
+  field_name = "one_minus_advected";
+  m = Metadata({Metadata::Cell, Metadata::Graphics, Metadata::Derived, Metadata::OneCopy});
+  pkg->AddField(field_name, m);
+
+  field_name = "one_minus_advected_sq";
+  pkg->AddField(field_name, m);
+
+  field_name = "one_minus_sqrt_one_minus_advected_sq";
+  pkg->AddField(field_name, m);
+
+  pkg->FillDerived = SquareIt;
+  pkg->CheckRefinement = CheckRefinement;
+  pkg->EstimateTimestep = EstimateTimestep;
+
+  return pkg;
+}
+
+int CheckRefinement(Container<Real>& rc) {
+  MeshBlock *pmb = rc.pmy_block;
+  // refine on advected, for example.  could also be a derived quantity
+  CellVariable<Real>& v = rc.Get("advected");
+  Real vmin = 1.0;
+  Real vmax = 0.0;
+  for (int k=0; k<pmb->ncells3; k++) {
+    for (int j=0; j<pmb->ncells2; j++) {
+      for (int i=0; i<pmb->ncells1; i++) {
+        vmin = (v(k,j,i) < vmin ? v(k,j,i) : vmin);
+        vmax = (v(k,j,i) > vmax ? v(k,j,i) : vmax);
+      }
+    }
+  }
+  auto pkg = pmb->packages["Advection"];
+  const auto& refine_tol = pkg->Param<Real>("refine_tol");
+  const auto& derefine_tol = pkg->Param<Real>("derefine_tol");
+
+  if (vmax > refine_tol && vmin < derefine_tol) return 1;
```

Comment:
Should we be using named constants or an enum for AMR flags such as REFINE, DEREFINE, NONE instead of 1, -1, and 0 respectively?

Reply:
Good idea.  Added enum class AmrTag : int {derefine=-1, same=0, refine=1}.  So now instead of return 1, it's return AmrTag::refine.  Better?